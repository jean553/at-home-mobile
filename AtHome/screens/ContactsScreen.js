import React from 'react';

import {
    Container,
    Text,
} from 'native-base';

import {
    Font,
    Contacts,
} from 'expo';

import {
    StyleSheet,
    PermissionsAndroid,
    BackHandler,
    FlatList,
    TouchableOpacity,
} from 'react-native';

export default class ContactsScreen extends React.Component {

    static navigationOptions = { title: 'I want to notify...' };

    /**
     *
     */
    constructor(props) {

        super(props);

        this.state = {
            fontsLoaded: false,
            contactsLoaded: false,
            contacts: [],
        };

        this.loadContacts = this.loadContacts.bind(this);
        this.openAddressPage = this.openAddressPage.bind(this);

        this.loadContacts();
    }

    /**
     * Asks for contacts access and loads the contacts if access granted. Closes the app if access denied.
     */
    async loadContacts() {

        const readContactsPermissionGranted = await Expo.Permissions.askAsync(Expo.Permissions.CONTACTS);

        if (readContactsPermissionGranted === 'granted') {
            BackHandler.exitApp();
            return;
        }

        await Contacts.getContactsAsync().then((contacts) => {
            this.setState({
                contacts: contacts.data,
                contactsLoaded: true,
            })
        });
    }

    /**
     * Redirects to the address selection page when a contact is clicked.
     */
    openAddressPage(phoneNumber) {

        const { navigate } = this.props.navigation;
        navigate("Address", { phoneNumber: phoneNumber });
    }

    /**
     * Mounts the TTF font files asynchronously,
     * this is required by native-base;
     * the page can be displayed after this operation only
     */
    async componentDidMount() {

        await Expo.Font.loadAsync({
            Roboto: require("native-base/Fonts/Roboto.ttf"),
            Roboto_medium: require("native-base/Fonts/Roboto_medium.ttf"),
            Ionicons: require("@expo/vector-icons/fonts/Ionicons.ttf"),
            Axiforma: require("../assets/fonts/axiforma/kastelov_-_axiforma_bold-webfont.ttf"),
        });

        this.setState({ fontsLoaded: true });
    }

    /**
     *
     */
    render() {

        if (
            this.state.fontsLoaded === false ||
            this.state.contactsLoaded === false
        ) {
            return null;
        }

        return (
            <Container>
                <FlatList
                    data={this.state.contacts}
                    renderItem={(contact) =>
                        <TouchableOpacity
                            key={contact.id}
                            onPress={() => {this.openAddressPage(contact.item.phoneNumbers[0].number)}}
                        >
                            <Text style={styles.contact}>{contact.item.firstName}</Text>
                        </TouchableOpacity>
                    }
                    keyExtractor={(contact, index) => contact.id}
                />
            </Container>
        );
    }
}

const styles = StyleSheet.create({
    contact: { padding: 10 }
});
