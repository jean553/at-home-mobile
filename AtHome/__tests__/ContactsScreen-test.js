/**
 * @jest-environment jsdom
 */

import ContactsScreen from '../screens/ContactsScreen';

import React from 'react';

import renderer from 'react-test-renderer';

import { shallow } from 'enzyme';

test('app starts correctly', () => {

    const component = shallow(<ContactsScreen />);
    component.setState({ componentIsReady: true });

    expect(component).toMatchSnapshot();
});
