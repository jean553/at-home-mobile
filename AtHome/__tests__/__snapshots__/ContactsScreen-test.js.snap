// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app starts correctly 1`] = `
<Styled(Container)>
  <FlatList
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Styled(Container)>
`;
